File: frontend/src/features/planning/planningBrief.js
/** 원자료와 별도로 보관하는 사용자 사업 여건. 지역별 초안과 생성 당시 조건은 구분합니다. */
export function emptyPlanningBrief(regionCode) {
  // API 스키마와 같은 기본 형태를 먼저 만들면, 미정 값도 0이나 빈 사실로 오해되지 않습니다.
  return { version: 1, region_code: regionCode, budget_status: 'unknown', budget_min_krw: null,
    budget_max_krw: null, budget_hard_limit: false, visitor_target_pct: null, spending_target_pct: null, schedule_status: 'unknown', start_date: null,
    end_date: null, resources_status: 'unknown', resources_confirmed: '', resources_possible: '',
    constraints_status: 'unknown', hard_constraints: '',
    preferences: '', field_context: '', references: [] }
}

export const BUSINESS_DIRECTIONS = [['auto', '사례 기반 추천'], ['spend_conversion', '지역 소비·환급'], ['stay_conversion', '숙박·체류'], ['night_time_experience', '야간 관광'], ['return_visit', '재방문·관광주민증']]
export const EXCLUDED_OPERATIONS = [['night_time_experience', '야간 운영 제외'], ['spend_conversion', '환급·소비지원 방식 제외']]
export function threeMonthSchedule(month) {
  if (!month) return { schedule_status: 'unknown', start_date: null, end_date: null }
  const [year, value] = month.split('-').map(Number)
  const end = new Date(Date.UTC(year, value + 2, 0)).toISOString().slice(0, 10)
  return { schedule_status: 'fixed', start_date: `${month}-01`, end_date: end }
}
export const RESOURCE_OPTIONS = [['information_center', '관광안내소'], ['merchants', '상인회·지역 상점'], ['lodging', '숙박업체'], ['events', '기존 행사'], ['cultural_spaces', '문화·체험 공간'], ['promotion', '지역 홍보 채널']]
export const CONTEXT_OPTIONS = [['families', '가족 방문객 중심'], ['young_adults', '청년 방문객 중심'], ['weekend', '주말 방문 연계'], ['weekdays', '평일 방문 확대'], ['event_link', '기존 행사 연계'], ['experience', '지역 체험 연계']]
export const optionLabels = (options, selected = []) => options.filter(([key]) => selected.includes(key)).map(([, label]) => label).join(' · ')
export function nextThreeMonthSchedule(now = new Date()) {
  const korea = new Date(now.getTime() + 9 * 60 * 60 * 1000)
  const offset = korea.getUTCDate() <= 15 ? 1 : 2
  const start = new Date(Date.UTC(korea.getUTCFullYear(), korea.getUTCMonth() + offset, 1))
  return threeMonthSchedule(start.toISOString().slice(0, 7))
}
export function simplifiedDraft(code) {
  const prior = readPlanningDraft(code)
  return { ...emptyPlanningBrief(code), input_profile: 'guided_v2', business_direction: prior.business_direction || 'auto', excluded_operations: [],
    budget_status: prior.budget_max_krw ? 'indicative' : 'unknown', budget_max_krw: prior.budget_max_krw,
    ...nextThreeMonthSchedule(),
    resource_options: RESOURCE_OPTIONS.filter(([key]) => prior.resource_options?.includes(key)).map(([key]) => key),
    context_options: CONTEXT_OPTIONS.filter(([key]) => prior.context_options?.includes(key)).map(([key]) => key) }
}

// 자유 입력과 첨부 본문은 같은 LLM 문맥을 사용하므로 합계도 제한합니다.
// 개별 칸 제한만 두면 최대 약 2만 5천 자가 되어 지역 근거가 긴 경우 생성을 시작한 뒤 실패할 수 있습니다.
export const PLANNING_CONTEXT_MAX_CHARS = 6000
const PLANNING_TEXT_FIELDS = ['resources_confirmed', 'resources_possible', 'hard_constraints', 'preferences', 'field_context']
export function planningContextCharCount(brief) {
  const fields = PLANNING_TEXT_FIELDS.reduce((total, field) => total + String(brief?.[field] || '').length, 0)
  const references = (brief?.references || []).reduce((total, reference) => total + String(reference?.text || '').length, 0)
  return fields + references
}

// 지역별 키를 사용해 강남구 초안과 다른 시군구 초안이 서로 덮어쓰지 않게 합니다.
const key = (code) => `tour-insight-planning-brief-${code}`
export function readPlanningDraft(code) {
  try {
    const saved = JSON.parse(window.localStorage.getItem(`${key(code)}-v2`) || window.localStorage.getItem(key(code)) || 'null')
    if (saved?.brief?.region_code === code && saved.brief.version === 1 && Array.isArray(saved.brief.references)) {
      // 이전 초안에는 상태 필드가 없으므로, 이미 적은 내용이 사라지지 않게 한 번만 보정합니다.
      const legacy = saved.brief
      return {
        ...emptyPlanningBrief(code), ...legacy,
        resources_status: legacy.resources_status || (legacy.resources_confirmed || legacy.resources_possible ? 'known' : 'unknown'),
        constraints_status: legacy.constraints_status || (legacy.hard_constraints ? 'known' : 'unknown'),
      }
    }
  } catch { /* 손상된 초안은 다른 지역 정보로 대체하지 않고 새로 시작합니다. */ }
  return emptyPlanningBrief(code)
}
export function savePlanningDraft(brief) {
  // 첨부 본문은 생성 요청에만 쓰고 브라우저 초안에는 남기지 않습니다.
  const draft = { ...brief, references: [] }
  window.localStorage.setItem(brief.input_profile === 'guided_v2' ? `${key(brief.region_code)}-v2` : key(brief.region_code), JSON.stringify({ brief: draft, saved_at: new Date().toISOString() }))
}
export function validatePlanningBrief(brief) {
  if (brief.input_profile === 'guided_v1') {
    if ((brief.excluded_operations || []).includes(brief.business_direction)) return '사업 방향과 제외 조건이 충돌합니다. 방향 또는 제외 조건을 바꿔 주세요.'
    if (brief.start_date && brief.start_date.slice(0, 7) < new Date().toISOString().slice(0, 7)) return '시작 월은 이번 달 이후로 선택해 주세요.'
  }
  const visitor = brief.visitor_target_pct
  const spending = brief.spending_target_pct
  if ((visitor != null) !== (spending != null)) return '방문자와 관광소비 목표율을 둘 다 입력하거나 모두 비워 주세요.'
  if (visitor != null && (!Number.isFinite(visitor) || visitor < 0 || visitor > 20 || !Number.isFinite(spending) || spending < 0 || spending > 30)) return '방문자 목표율은 0~20%, 관광소비 목표율은 0~30%로 입력해 주세요.'
  // 서버 Pydantic 검증 전에 같은 규칙을 한 번 적용해 사용자가 바로 오류를 알 수 있게 합니다.
  if (brief.budget_status !== 'unknown') {
    if (!Number.isSafeInteger(brief.budget_max_krw) || brief.budget_max_krw <= 0 || brief.budget_max_krw > 1e12) return '예산은 1원 이상 1조 원 이하의 정수로 입력해 주세요.'
    if (brief.budget_min_krw !== null && (!Number.isSafeInteger(brief.budget_min_krw) || brief.budget_min_krw <= 0 || brief.budget_min_krw > brief.budget_max_krw)) return '최소 예산은 최대 예산 이하로 입력해 주세요.'
  }
  if (brief.schedule_status !== 'unknown') {
    if (!brief.start_date || !brief.end_date) return '사업 시작일과 종료일을 모두 입력해 주세요.'
    if (brief.end_date < brief.start_date) return '종료일은 시작일보다 빠를 수 없습니다.'
  }
  if (planningContextCharCount(brief) > PLANNING_CONTEXT_MAX_CHARS) return `현장 정보·조건·선호·참고문서 본문은 합계 ${PLANNING_CONTEXT_MAX_CHARS.toLocaleString('ko-KR')}자 이하로 줄여 주세요.`
  return ''
}
export function briefBudget(brief) {
  // 입력 숫자를 요약 카드에서 읽기 좋은 ‘원’ 단위 문자열로 바꿉니다.
  if (!brief || brief.budget_status === 'unknown') return '미정 · AI가 규모 제안'
  if (!Number.isSafeInteger(brief.budget_max_krw) || brief.budget_max_krw <= 0) return '금액 입력 필요'
  const amount = (n) => `${Number(n).toLocaleString('ko-KR')}원`
  return `${brief.budget_min_krw ? amount(brief.budget_min_krw) + ' ~ ' : ''}${amount(brief.budget_max_krw)}${brief.budget_hard_limit ? ' 이내' : ''}`
}
export function briefPeriod(brief) {
  if (brief?.input_profile === 'guided_v1' && brief.schedule_status === 'unknown') return '이번 달부터 3개월';
  // 날짜 둘 중 하나만 입력된 상태는 저장하면 안 되므로 요약에도 ‘입력 필요’로 표시합니다.
  if (!brief || brief.schedule_status === 'unknown') return '미정 · AI가 일정 제안'
  return brief.start_date && brief.end_date ? `${brief.start_date} ~ ${brief.end_date}` : '날짜 입력 필요'
}

File: ai_server/app/planning_brief.py
"""사업 여건 계약. 사용자 입력은 관측 사실/공용 RAG와 분리해 보관합니다."""
from __future__ import annotations

from datetime import date, datetime, timedelta, timezone
import calendar
from hashlib import sha256
from io import BytesIO
import json
from pathlib import Path
from typing import Literal
from zipfile import ZipFile, BadZipFile

from docx import Document
from openpyxl import load_workbook
from pypdf import PdfReader
from pydantic import BaseModel, ConfigDict, Field, model_validator


PLANNING_CONTEXT_MAX_CHARS = 6000
PLANNING_TEXT_FIELDS = ('resources_confirmed', 'resources_possible', 'hard_constraints', 'preferences', 'field_context')


def planning_context_char_count(brief: 'PlanningBrief | dict | None') -> int:
    """LLM에 전달되는 사용자 자유 입력과 첨부 본문의 전체 문자 수입니다."""
    if brief is None:
        return 0
    value = brief if isinstance(brief, dict) else brief.model_dump()
    field_chars = sum(len(str(value.get(field) or '')) for field in PLANNING_TEXT_FIELDS)
    reference_chars = sum(len(str(row.get('text') or '')) for row in value.get('references') or [])
    return field_chars + reference_chars


RESOURCE_OPTIONS = {'information_center': '관광안내소', 'merchants': '상인회·지역 상점', 'lodging': '숙박업체', 'events': '기존 행사', 'cultural_spaces': '문화·체험 공간', 'promotion': '지역 홍보 채널'}
CONTEXT_OPTIONS = {'families': '가족 방문객 중심', 'young_adults': '청년 방문객 중심', 'weekend': '주말 방문 연계', 'weekdays': '평일 방문 확대', 'event_link': '기존 행사 연계', 'experience': '지역 체험 연계'}


def next_three_months(as_of_date: date | None = None) -> tuple[date, date]:
    """한국 시간 15일까지 다음 달, 16일부터 다다음 달 시작의 3개월입니다."""
    today = as_of_date or datetime.now(timezone(timedelta(hours=9))).date()
    offset = 1 if today.day <= 15 else 2
    start_index = today.year * 12 + today.month - 1 + offset
    year, month = divmod(start_index, 12)
    end_year, end_month = divmod(start_index + 2, 12)
    return date(year, month + 1, 1), date(end_year, end_month + 1, calendar.monthrange(end_year, end_month + 1)[1])


def resolve_new_planning_brief(brief: 'PlanningBrief | None', *, as_of_date: date | None = None) -> 'PlanningBrief | None':
    # 신규 요청에만 호출합니다. 저장 보고서/재시작 작업을 읽을 때 날짜를 바꾸지 않습니다.
    if brief is None or brief.input_profile != 'guided_v2':
        return brief
    start, end = next_three_months(as_of_date)
    return PlanningBrief.model_validate({**brief.model_dump(), 'schedule_status': 'fixed', 'start_date': start, 'end_date': end})


class BriefReference(BaseModel):
    model_config = ConfigDict(extra='forbid', str_strip_whitespace=True)
    name: str = Field(min_length=1, max_length=150)
    text: str = Field(min_length=1, max_length=6000)


class PlanningBrief(BaseModel):
    """정해지지 않은 값은 null/unknown으로 유지합니다. 0원이나 확정으로 바꾸지 않습니다."""
    model_config = ConfigDict(extra='forbid', str_strip_whitespace=True)
    version: Literal[1] = 1
    input_profile: Literal['legacy', 'guided_v1', 'guided_v2'] = 'legacy'
    business_direction: Literal['auto', 'spend_conversion', 'stay_conversion', 'night_time_experience', 'return_visit'] = 'auto'
    resource_options: list[Literal['information_center', 'merchants', 'lodging', 'events', 'cultural_spaces', 'promotion']] = Field(default_factory=list, max_length=6)
    context_options: list[Literal['families', 'young_adults', 'weekend', 'weekdays', 'event_link', 'experience']] = Field(default_factory=list, max_length=6)
    excluded_operations: list[Literal['night_time_experience', 'spend_conversion']] = Field(default_factory=list, max_length=2)
    region_code: str = Field(pattern=r'^\d{2,5}$')
    budget_status: Literal['unknown', 'indicative', 'confirmed'] = 'unknown'
    budget_min_krw: int | None = Field(default=None, ge=1, le=1_000_000_000_000, strict=True)
    budget_max_krw: int | None = Field(default=None, ge=1, le=1_000_000_000_000, strict=True)
    budget_hard_limit: bool = False
    visitor_target_pct: float | None = Field(default=None, ge=0, le=20, allow_inf_nan=False)
    spending_target_pct: float | None = Field(default=None, ge=0, le=30, allow_inf_nan=False)
    schedule_status: Literal['unknown', 'flexible', 'fixed'] = 'unknown'
    start_date: date | None = None
    end_date: date | None = None
    # 빈 입력과 ‘알 수 없음’을 구분해 AI가 시설·인력이 없다고 단정하지 않게 합니다.
    resources_status: Literal['unknown', 'known'] = 'unknown'
    resources_confirmed: str = Field(default='', max_length=1500)
    resources_possible: str = Field(default='', max_length=1500)
    constraints_status: Literal['unknown', 'known'] = 'unknown'
    hard_constraints: str = Field(default='', max_length=2000)
    preferences: str = Field(default='', max_length=1000)
    field_context: str = Field(default='', max_length=2500)
    references: list[BriefReference] = Field(default_factory=list, max_length=3)

    @model_validator(mode='before')
    @classmethod
    def infer_legacy_input_status(cls, value):
        """상태 필드가 없던 기존 브라우저 초안도 내용을 잃지 않고 읽습니다."""
        if not isinstance(value, dict):
            return value
        data = dict(value)
        if 'resources_status' not in data:
            data['resources_status'] = 'known' if data.get('resources_confirmed') or data.get('resources_possible') else 'unknown'
        if 'constraints_status' not in data:
            data['constraints_status'] = 'known' if data.get('hard_constraints') else 'unknown'
        return data

    @model_validator(mode='after')
    def check_conditions(self):
        if self.input_profile == 'guided_v2':
            if self.excluded_operations or self.hard_constraints or self.preferences or self.resources_possible or self.references:
                raise ValueError('선택형 기획은 제공된 자원·현장 항목만 지원합니다.')
            if self.budget_hard_limit or self.budget_min_krw is not None or self.budget_status == 'confirmed':
                raise ValueError('선택형 기획은 참고 예산 총액만 지원합니다.')
            if self.visitor_target_pct is not None or self.spending_target_pct is not None:
                raise ValueError('목표 KPI는 생성 후 조정해 주세요.')
            self.resource_options = list(dict.fromkeys(self.resource_options))
            self.context_options = list(dict.fromkeys(self.context_options))
            # 자유 문구가 함께 전송되어도 선택 코드에서만 LLM 참고 문장을 구성합니다.
            self.resources_confirmed = ' · '.join(RESOURCE_OPTIONS[key] for key in self.resource_options)
            self.field_context = ' · '.join(CONTEXT_OPTIONS[key] for key in self.context_options)
            self.resources_status = 'known' if self.resource_options else 'unknown'
            self.constraints_status = 'unknown'
        if self.input_profile == 'guided_v1':
            import calendar
            if self.schedule_status == 'unknown':
                today = date.today()
                self.start_date = today.replace(day=1)
                end_index = today.year * 12 + today.month - 1 + 2
                end_year, end_month = divmod(end_index, 12)
                self.end_date = date(end_year, end_month + 1, calendar.monthrange(end_year, end_month + 1)[1])
                self.schedule_status = 'fixed'
            if self.business_direction in self.excluded_operations:
                raise ValueError('사업 방향과 제외 조건이 충돌합니다.')
            if self.budget_hard_limit or self.budget_min_krw is not None or self.budget_status == 'confirmed':
                raise ValueError('간편 기획은 참고 예산 총액만 지원합니다.')
            if self.visitor_target_pct is not None or self.spending_target_pct is not None or self.references:
                raise ValueError('간편 기획의 목표는 생성 후 조정하며 첨부자료는 지원하지 않습니다.')
            if self.preferences or self.resources_possible or self.hard_constraints:
                raise ValueError('간편 기획은 선택형 제외 조건과 현장 메모만 지원합니다.')
            if len(self.resources_confirmed) > 300 or len(self.field_context) > 500:
                raise ValueError('활용 자원은 300자, 현장 메모는 500자 이하로 입력하세요.')
            if self.start_date and self.end_date:
                index = self.start_date.year * 12 + self.start_date.month - 1 + 2
                year, month = divmod(index, 12)
                expected = date(year, month + 1, calendar.monthrange(year, month + 1)[1])
                if self.start_date.day != 1 or self.end_date != expected or self.start_date.strftime('%Y%m') < date.today().strftime('%Y%m'):
                    raise ValueError('이번 달 이후 시작 월부터 3개월만 지원합니다.')
        if (self.visitor_target_pct is None) != (self.spending_target_pct is None):
            raise ValueError('방문자와 관광소비 목표율은 둘 다 입력하거나 모두 미정으로 두세요.')
        if self.budget_status == 'unknown':
            if self.budget_min_krw is not None or self.budget_max_krw is not None or self.budget_hard_limit:
                raise ValueError('예산 미정 상태에서는 금액과 상한을 지정할 수 없습니다.')
        elif self.budget_max_krw is None:
            raise ValueError('예산 금액 또는 범위의 최대 금액을 입력해 주세요.')
        if self.budget_min_krw and self.budget_max_krw and self.budget_min_krw > self.budget_max_krw:
            raise ValueError('최소 예산은 최대 예산보다 클 수 없습니다.')
        if self.schedule_status == 'unknown' and (self.start_date or self.end_date):
            raise ValueError('일정 미정 상태에서는 날짜를 지정할 수 없습니다.')
        if self.schedule_status != 'unknown' and not (self.start_date and self.end_date):
            raise ValueError('사업 시작일과 종료일을 모두 입력해 주세요.')
        if self.start_date and self.end_date and self.start_date > self.end_date:
            raise ValueError('사업 종료일은 시작일보다 빠를 수 없습니다.')
        if self.resources_status == 'unknown' and (self.resources_confirmed or self.resources_possible):
            raise ValueError('시설·인력 미정 상태에서는 내용을 입력할 수 없습니다.')
        if self.constraints_status == 'unknown' and self.hard_constraints:
            raise ValueError('필수 조건 미정 상태에서는 내용을 입력할 수 없습니다.')
        if planning_context_char_count(self) > PLANNING_CONTEXT_MAX_CHARS:
            raise ValueError(
                f'현장 정보·조건·선호·참고문서 본문은 합계 {PLANNING_CONTEXT_MAX_CHARS:,}자 이하로 줄여 주세요.'
            )
        return self


def brief_fingerprint(brief: dict | None) -> str:
    if not brief:
        return ''
    return sha256(json.dumps(brief, ensure_ascii=False, sort_keys=True).encode()).hexdigest()


def without_reference_text(brief: PlanningBrief | None) -> PlanningBrief | None:
    """생성 뒤 보관·응답되는 조건에서는 첨부 본문을 제거합니다."""
    return brief.model_copy(update={'references': []}) if brief else None


def brief_summary(brief: dict | None) -> str:
    """화면/다운로드용 짧은 조건 요약. 출처 수치가 아닌 사용자 설정임을 명시합니다."""
    if not brief:
        return ''
    amount = brief.get('budget_max_krw')
    budget = f"{amount:,}원" if amount else '미정'
    budget += ' (상한)' if brief.get('budget_hard_limit') else ''
    period = f"{brief['start_date']} ~ {brief['end_date']}" if brief.get('start_date') and brief.get('end_date') else '미정'
    return f"사용자 입력 조건 | 예산 {budget} · 일정 {period}"


def extract_brief_reference(filename: str, content: bytes) -> dict:
    """첨부 문서의 필요한 텍스트만 메모리에서 읽고, 원본은 저장하지 않습니다."""
    if not content or len(content) > 2_000_000:
        raise ValueError('파일은 2MB 이하로 첨부해 주세요.')
    name = Path(filename.replace('\\', '/')).name[:150]
    suffix = Path(name).suffix.lower()
    if suffix in ('.txt', '.md'):
        try:
            text = content.decode('utf-8-sig')
        except UnicodeDecodeError:
            text = content.decode('cp949')
    elif suffix in ('.docx', '.hwpx', '.xlsx'):
        try:
            with ZipFile(BytesIO(content)) as archive:
                files = archive.infolist()
                if len(files) > 500 or sum(item.file_size for item in files) > 12_000_000:
                    raise ValueError('압축 해제 크기가 너무 큰 문서는 첨부할 수 없습니다.')
            if suffix == '.docx':
                document = Document(BytesIO(content))
                text = '\n'.join([p.text for p in document.paragraphs] + [
                    ' | '.join(c.text for c in row.cells) for t in document.tables for row in t.rows
                ])
            elif suffix == '.hwpx':
                text = _extract_hwpx_text(BytesIO(content))
            else:
                text = _extract_xlsx_text(BytesIO(content))
        except (BadZipFile, KeyError) as exc:
            raise ValueError('올바른 Word·한글(HWPX)·Excel 파일을 첨부해 주세요.') from exc
    elif suffix == '.pdf':
        try:
            reader = PdfReader(BytesIO(content))
            if reader.is_encrypted:
                raise ValueError('암호가 설정된 PDF는 텍스트를 읽을 수 없습니다.')
            text = '\n'.join(page.extract_text() or '' for page in reader.pages[:40])
        except ValueError:
            raise
        except Exception as exc:
            raise ValueError('텍스트를 읽을 수 있는 PDF를 첨부해 주세요.') from exc
    else:
        raise ValueError('한글(HWPX), Word(DOCX), PDF, TXT, Excel(XLSX) 파일을 첨부해 주세요. HWP는 HWPX 또는 PDF로 저장해 주세요.')
    text = text.replace('\x00', '').strip()
    if not text:
        raise ValueError('문서에서 텍스트를 찾지 못했습니다.')
    if len(text) > 6000:
        raise ValueError('문서 내용이 6,000자를 넘습니다. 필요한 부분만 별도 문서로 첨부해 주세요.')
    return BriefReference(name=name, text=text).model_dump()


def _extract_hwpx_text(buffer: BytesIO) -> str:
    """HWPX는 ZIP 안 XML 문서라 외부 변환기 없이 본문 텍스트만 읽을 수 있습니다."""
    import xml.etree.ElementTree as etree

    with ZipFile(buffer) as archive:
        sections = sorted(name for name in archive.namelist() if name.startswith('Contents/section') and name.endswith('.xml'))
        if not sections:
            raise ValueError('본문이 없는 HWPX 파일입니다.')
        return '\n'.join(''.join(element.itertext()) for name in sections for element in [etree.fromstring(archive.read(name))])


def _extract_xlsx_text(buffer: BytesIO) -> str:
    """Excel은 첫 10개 시트·각 500행만 읽어, 참고자료 처리 시간을 제한합니다."""
    workbook = load_workbook(buffer, read_only=True, data_only=True)
    rows: list[str] = []
    try:
        for worksheet in workbook.worksheets[:10]:
            rows.append(f'[{worksheet.title}]')
            for row in worksheet.iter_rows(max_row=500, values_only=True):
                values = [str(value).strip() for value in row[:50] if value is not None and str(value).strip()]
                if values:
                    rows.append(' | '.join(values))
    finally:
        workbook.close()
    return '\n'.join(rows)
